src/compute-client/src/controller/instance.rs

```diff
@@ -95,15 +95,48 @@ pub(super) enum SubscribeTargetError {
 pub(super) struct Instance<T> {
     /// Build info for spawning replicas
     build_info: &'static BuildInfo,
+    /// Whether instance initialization has been completed.
+    initialized: bool,
     /// The replicas of this compute instance.
     replicas: BTreeMap<ReplicaId, Replica<T>>,
-    /// Tracks expressed `since` and received `upper` frontiers for indexes and sinks.
+    /// Currently installed compute collections.
+    ///
+    /// New entries are added for all collections exported from dataflows created through
+    /// [`ActiveInstance::create_dataflows`].
+    ///
+    /// Entries are removed when two conditions are fulfilled:
+    ///  * The collection's read frontier has advanced to the empty frontier, implying that
+    ///    [`ActiveInstance::drop_collections`] was called.
+    ///  * All replicas have reported the empty frontier for the collection, implying that they
+    ///    have stopped reading from the collection's inputs.
+    ///
+    /// Only if both these conditions hold are is dropping a collection's state, and the associated
+    /// read holds on its inputs, sound.
     collections: BTreeMap<GlobalId, CollectionState<T>>,
     /// IDs of arranged log sources maintained by this compute instance.
     arranged_logs: BTreeMap<LogVariant, GlobalId>,
     /// Currently outstanding peeks.
+    ///
+    /// New entries are added for all peeks initiated through [`ActiveInstance::peek`].
+    ///
+    /// The entry for a peek is only removed once all replicas have responsed to the peek. This is
+    /// currently required to ensure all replicas have stopped reading from the peeked collection's
+    /// inputs before we allow them to compact. #16641 tracks changing this so we only have to wait
+    /// for the first peek response.
     peeks: BTreeMap<Uuid, PendingPeek<T>>,
     /// Currently in-progress subscribes.
+    ///
+    /// New entries are added for all subscribes exported from dataflows created through
+    /// [`ActiveInstance::create_dataflows`].
+    ///
+    /// The entry for a subscribe is removed once at least one replica has reported the subscribe
```

Comment:
This behavior is different from `peeks`, where we need to wait for everybody. Is #16641 meant to align `peeks` with the current behavior in `subscribers` or are we aiming for something else there?

Reply:
Yes, #16641 will make the behavior for peeks and subscribes match, in that we drop the tracking state as soon as we don't need it anymore to emit responses upstream. The only reason we keep peek state around longer currently is that we otherwise might release read holds on the peek input collections too soon and crash straggling replicas.